cli/commands/index.py

```diff
@@ -184,6 +190,87 @@ def get(
 
 
 @app.command()
-def replicate():
-    """Replicate an index from a production cluster to the rank cluster"""
-    raise NotImplementedError
+def replicate(
+    context: typer.Context,
+    source_index: str = typer.Option(
+        None,
+        help=(
+            "The name of the index to replicate. If an index is not provided, "
+            "you will be prompted to select one from the production cluster"
+        ),
+    ),
+    dest_index: str = typer.Option(
+        None,
+        help=(
+            "The name of the index to create in the rank cluster. If an index "
+            "is not provided, you will be prompted to select one from the rank "
+            "cluster. The default is to use the same name as the source index"
+        ),
+    ),
+):
+    """Reindex an index from a production cluster to the rank cluster"""
+    rank_client: Elasticsearch = context.meta["rank_client"]
+    search_templates = requests.get(
+        f"{ context.meta['catalogue_api_url']}/search-templates.json"
+    ).json()["templates"]
+
+    works = next(
+        template
+        for template in search_templates
+        if template["index"].startswith("works")
+    )
+
+    pipeline_date = re.search(
+        r"^works-indexed-(?P<date>\d{4}-\d{2}-\d{2}.?)", works["index"]
+    ).group("date")
+
+    secret_prefix = f"elasticsearch/pipeline_storage_{pipeline_date}/"
+
+    get_secret = functools.partial(aws.get_secret, context.meta["session"])
+
+    pipeline_password = get_secret(secret_prefix + "es_password")
+    pipeline_username = get_secret(secret_prefix + "es_username")
+    protocol = get_secret(secret_prefix + "protocol")
+    public_host = get_secret(secret_prefix + "public_host")
+    port = get_secret(secret_prefix + "port")
+    pipeline_host = f"{protocol}://{public_host}:{port}"
+
+    pipeline_client = elasticsearch.pipeline_client(
+        session=context.meta["session"],
+        pipeline_date=pipeline_date,
+    )
```

Comment:
is there a reason we can't use `pipeline_client` in `services/elasticsearch.py` to do this?

Reply:
The only reason for this big repeated block is that we need the host/credentials later on for

```json
"remote": {
    "host": pipeline_host,
    "username": pipeline_username,
    "password": pipeline_password,
},
```

and I don't think we can get those values back out of an existing client. It is ugly though...

Reply:
In that case maybe create a `get_cluster_credentials` method to be used in both places? And accept that it'd be used twice when calling `pipeline_client` here as well

Reply:
I've morphed `get_secret()`  into `get_secrets()`, which achieves roughly the same thing. I think there's probably a cleverer version of this which uses the `context.meta` somehow, but this works for now 🤷

Let me know what you think